### PR TITLE
Smarter handling if queue contains only "no-signal" ops.

### PR DESCRIPTION
**Pull request by bensander**
Commands can be submitted w/o a completion signal.
Now if queue contains only these "no-signal" ops and isEmpty()
is called, then we add a barrier and return false.  Subsequent
calls to isEmpty() will test the signal associated with the barrier
to see if the queue is empty.

### CODE REVIEW

File: lib/hsa/mcwamp_hsa.cpp
Comment (AlexVlx) [inline lib/hsa/mcwamp_hsa.cpp]: @bensander couldn't this be just be find_if with reverse iterators + the associated processing? Which is to say something along the lines of: 
```cpp
const auto oldest = find_if(
    asyncOps.crbegin(), asyncOps.crend(), [](... asyncOp) { return asyncOp != nullptr; });

if (oldest != asyncOps.crend()) { /* do the check for signal */ }
```
The loop is somewhat unnatural as is (it appears more customary to use while loops when counting down).

File: lib/hsa/mcwamp_hsa.cpp
Comment (AlexVlx) [inline lib/hsa/mcwamp_hsa.cpp]: Are we sure relaxed with no fence is what we want here? I believe that this is concurrently written by the CP, is it not?

**Review by AlexVlx — CHANGES_REQUESTED** [review]
In general LGTM, but could you please peek at the 2 items I mentioned? Thank you @bensander.

File: lib/hsa/mcwamp_hsa.cpp
Comment (bensander) [reply lib/hsa/mcwamp_hsa.cpp]: This is unchanged from code before, but probably should be changed.  It comes down to what effect is implied on the memory model if a user submits a command and then checks if the associated accelerator view is empty - probably would assume that the results of that command are visible.

File: lib/hsa/mcwamp_hsa.cpp
Comment (bensander) [reply lib/hsa/mcwamp_hsa.cpp]: yep, I think this can be refactored as you suggest, will submit a new one.  It started as a forward loop ; now that it runs backward all we really have to do is find the oldest op and see if it has a signal or not.

File: lib/hsa/mcwamp_hsa.cpp
Comment (AlexVlx) [inline lib/hsa/mcwamp_hsa.cpp]: Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

File: lib/hsa/mcwamp_hsa.cpp
Comment (AlexVlx) [inline lib/hsa/mcwamp_hsa.cpp:4979]: Unless I am mistaken, this is deprecated, with `hsa_signal_wait_scacquire` being the current / preferred form.

File: lib/hsa/mcwamp_hsa.cpp
Comment (AlexVlx) [inline lib/hsa/mcwamp_hsa.cpp]: Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

File: lib/hsa/mcwamp_hsa.cpp
Comment (AlexVlx) [inline lib/hsa/mcwamp_hsa.cpp]: Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

File: lib/hsa/mcwamp_hsa.cpp
Comment (AlexVlx) [inline lib/hsa/mcwamp_hsa.cpp:4702]: Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

File: lib/hsa/mcwamp_hsa.cpp
Comment (AlexVlx) [inline lib/hsa/mcwamp_hsa.cpp]: Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

File: lib/hsa/mcwamp_hsa.cpp
Comment (AlexVlx) [inline lib/hsa/mcwamp_hsa.cpp:5394]: Should we put this in a loop, since it's both a relaxed wait and it is legal for a signal to spuriously return?

File: lib/hsa/mcwamp_hsa.cpp
Comment (AlexVlx) [inline lib/hsa/mcwamp_hsa.cpp:5381]: Should either consider using screlease or a release fence here, to ensure no reordering / fully observable precursors as far as the CP is concerned?

File: lib/hsa/mcwamp_hsa.cpp
Comment (AlexVlx) [inline lib/hsa/mcwamp_hsa.cpp:4438]: Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

File: lib/hsa/mcwamp_hsa.cpp
Comment (AlexVlx) [inline lib/hsa/mcwamp_hsa.cpp]: Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

File: lib/hsa/mcwamp_hsa.cpp
Comment (AlexVlx) [inline lib/hsa/mcwamp_hsa.cpp]: Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

**Review by AlexVlx — COMMENTED** [review]
Thank you for the changes, LGTM. Now, since you went beyond the call of duty, and touched a few other bits, there are a few other (possibly pedantic :) ) concerns, but it is your call if you want to address them or not.

File: lib/hsa/mcwamp_hsa.cpp
Comment (bensander) [reply lib/hsa/mcwamp_hsa.cpp:5394]: Most of these are already present in the code and unrelated to the new functionality, lets treat as a new PR.

File: lib/hsa/mcwamp_hsa.cpp
Comment (bensander) [reply lib/hsa/mcwamp_hsa.cpp:5381]: These are resetting signal value so relaxed seems fine.     CP will store-release to set to 0 ; any consumers wait for the 0 value so that is important point to release.

File: lib/hsa/mcwamp_hsa.cpp
Comment (scchan) [inline lib/hsa/mcwamp_hsa.cpp:1505]: what's the reason for enqueueing a marker?

**Review by AlexVlx — APPROVED** [review]

File: lib/hsa/mcwamp_hsa.cpp
Comment (bensander) [reply lib/hsa/mcwamp_hsa.cpp:1505]: If last command had no completion signal then we have no way to tell if he queue is empty or not.  So we enque a marker here.  If app calls is empty again the marker will be oldest and we can get a valid read on the status.

**Pull request merged**
